File: DSA/Ayush_Mishra_590014844/Day_51_Question_1.c
#include <stdio.h>
#include <string.h>

int longestSubstring(char *str, int k) {
    int freq[256] = {0}, uniqueCount = 0, maxLen = 0;
    int left = 0, right = 0, n = strlen(str);

    while (right < n) {
        if (freq[str[right]] == 0)
            uniqueCount++;
        freq[str[right]]++;
        while (uniqueCount > k) {
            freq[str[left]]--;
            if (freq[str[left]] == 0)
                uniqueCount--;
            left++;
        }
        if (uniqueCount == k && (right - left + 1) > maxLen)
            maxLen = right - left + 1;
        right++;
    }
    return maxLen;
}

int main() {
    char str[1000];
    int k;
    scanf("%s", str);
    scanf("%d", &k);
    printf("%d\n", longestSubstring(str, k));
    return 0;
}

File: DSA/Ayush_Mishra_590014844/Day_51_Question_1.py
def longest_substring_k_unique(s, k):
    freq = {}
    left = 0
    max_len = 0

    for right in range(len(s)):
        freq[s[right]] = freq.get(s[right], 0) + 1
        while len(freq) > k:
            freq[s[left]] -= 1
            if freq[s[left]] == 0:
                del freq[s[left]]
            left += 1
        if len(freq) == k:
            max_len = max(max_len, right - left + 1)
    return max_len

s = input().strip()
k = int(input())
print(longest_substring_k_unique(s, k))
